### PR TITLE
MLDB-1315 row as table expression

### PR DESCRIPTION
**Pull request by jeremybarnes**
This allows rows to be used in table expressions, allowing rewriting and filtering of rows and gathering using joins.

### CODE REVIEW

File: container_files/assets/www/doc/builtin/sql/FromExpression.md [hunk 1]
```diff
@@ -78,3 +78,54 @@ SELECT x.* FROM merge(dataset1, dataset2, dataset3) AS x
...
+
+    }
+}
+
+SELECT row_transform({input: {x: 1, y: 2, z: "three"}})[output] AS *
+
+{ Z: "three" }
+```
+
+It can also be used to join a row against another table, in order to
```
Comment (nicolaskruchten) [inline container_files/assets/www/doc/builtin/sql/FromExpression.md]: I'm not at all sure I understand the use-case here. Can you describe a real-world scenario?

File: container_files/assets/www/doc/builtin/sql/FromExpression.md [hunk 2]
```diff
@@ -78,3 +78,54 @@ SELECT x.* FROM merge(dataset1, dataset2, dataset3) AS x
...
+
+As an example, the following `sql.query` object would strip
+out any numeric-valued columns and uppercase all names from a
+passed in row:
+
+```
+POST /v1/functions/row_transform {
+    type: 'sql.query',
+    params: {
+        query: 'SELECT upper(rowName()) AS column, value FROM ROW TABLE ($input) WHERE CAST (value AS NUMBER) IS NULL',
```
Comment (nicolaskruchten) [inline container_files/assets/www/doc/builtin/sql/FromExpression.md]: I would prefer syntax that looks more like a function-call like `sample` or `transpose` which we already support. Perhaps `row_table(...)` ?

File: container_files/assets/www/doc/builtin/sql/FromExpression.md (hunk 1 above)
Comment (nicolaskruchten) [inline container_files/assets/www/doc/builtin/sql/FromExpression.md]: OK, I see the scenario in the JIRA issue.

File: container_files/assets/www/doc/builtin/sql/FromExpression.md (hunk 2 above)
Comment (nicolaskruchten) [inline container_files/assets/www/doc/builtin/sql/FromExpression.md]: The fact that a `ROW TABLE` has a `rowName()` function seems very counter-intuitive, which leads me to think that the naming is off.

**Comment by nicolaskruchten** [comment]
-1 while we consider the DOD.

File: container_files/assets/www/doc/builtin/sql/FromExpression.md (hunk 2 above)
Comment (jeremybarnes) [inline container_files/assets/www/doc/builtin/sql/FromExpression.md]: All datasets have a rowName() function.  Would it make more sense that the returned dataset had a column called 'column'?  That way we wouldn't need to use the rowName() in the query.

**Comment by nicolaskruchten** [comment]
Sorry about the multiple name changes here but I just realized that we don't use the word "table" in MLDB, we try to use "dataset" consistently. So we should probably change this to `row_dataset`.

**Comment by nicolaskruchten** [comment]
+1 to merge after `row_table` -> `row_dataset`

**Pull request merged**